src/frequenz/sdk/microgrid/component_graph.py

```diff
@@ -853,10 +853,31 @@ def find_first_descendant_component(
             ValueError: When the root component is not found in the component
                 graph or when no component is found in the given categories.
         """
-        root_component = next(
-            (comp for comp in self.components(component_categories={root_category})),
-            None,
-        )
+        if root_category == ComponentCategory.GRID:
+            root_component = next(
+                (
+                    comp
+                    for comp in self.components(component_categories={root_category})
+                ),
+                None,
+            )
+        else:
+            grid_component = next(
+                (
+                    comp
+                    for comp in self.components(
+                        component_categories={ComponentCategory.GRID}
+                    )
+                ),
+                None,
+            )
+            if grid_component is None:
+                raise ValueError("Grid component not found in component graph!")
+
+            root_component = self.find_first_descendant_component(
+                root_category=ComponentCategory.GRID,
+                descendant_categories=[root_category],
+            )
```

Comment:
Alternative with less duplication for finding the grid component:

```suggestion
        grid_component = next(
            (
                comp
                for comp in self.components(
                    component_categories={ComponentCategory.GRID}
                )
            ),
            None,
        )
        if root_category == ComponentCategory.GRID:
            root_component = grid_component
        else:
            if grid_component is None:
                raise InvalidGraphError("Grid component not found in component graph")

            root_component = self.find_first_descendant_component(
                root_category=ComponentCategory.GRID,
                descendant_categories=[root_category],
            )
```

Reply:
dropped this section.